docs/madr/decisions/038-meshmetric-profiles.md

```diff
@@ -0,0 +1,352 @@
+
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#general (TCP related)
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#tcp-statistics
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/stats.html#tcp-statistics
+
+##### TLS
+
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#tls-statistics
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/tls_inspector.html#statistics
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/stats.html#tls-statistics
+
+##### Fault injection
+
+All stats from: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter.html#statistics
+
+##### RBAC
+
+All stats from:
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rbac_filter.html#statistics 
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/rbac_filter.html#statistics
+
+##### EnvoyServer
+
+All stats from: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/statistics.html#statistics
+
+##### AccessLog
+
+All stats from: https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/stats.html#statistics
+
+##### LoadBalancer
+
+All stats from: https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#load-balancer-statistics
+
+##### CircuitBreaker
+
+For circuit breaker we use both circuit breaker and outlier detection so we could combine these:
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#circuit-breakers-statistics
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats.html#outlier-detection-statistics
+
+##### Tracing
+
+- https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/stats.html#tracing-statistics
+
+#### Schema
+
+I suggest changing the current schema of
+
+```yaml
+sidecar:
+  usedOnly: true # true or false
+  profile: minimal # one of minimal, default, full
+  regex: http2_act.* # only profile or regex can be defined
+```
+
+to
+
+```yaml
+sidecar:
+  usedOnly: true # true or false
+  profile:
+    name: default # one of `nothing`, `minimal`, `default`, `comprehensive`, `all`
+    exclude: regex2.* # first exclude
+    include: regex1.* # then include (include takes over)
+```
+
+#### Implementation
+
+Just like we [mutate responses for metrics hijacker](https://github.com/kumahq/kuma/blob/d6c9ce64ac5e7ba1f5dbb9fb410e7d9410b67815/app/kuma-dp/pkg/dataplane/metrics/server.go#L348)
+we can add a filter mutator to reduce the number of metrics (same thing for [OTEL](https://github.com/kumahq/kuma/blob/d6c9ce64ac5e7ba1f5dbb9fb410e7d9410b67815/app/kuma-dp/pkg/dataplane/metrics/metrics_producer.go#L106)).
```

Comment:
Can we double check if we won't get a subsential memory usage reduction when we do this directly in Envoy?

I know that it's not dynamic then, because it's configured in bootstrap, but it might be worth checking if this is neglegable difference or not.

Reply:
> Can we double check if we won't get a subsential memory usage reduction when we do this directly in Envoy?

Yeah if we do this in Envoy then we'll reduce the memory:
- Envoy could be more memory efficient at filtering stuff
- 1 smaller response to our hijacker
- No need to allocate anything to process the response

I'll add the two approaches with pros and cons. I think we can start with doing this in Envoy (we pretty much get it for free with `filter`) and in the future if we need more processing we can rewrite that (it's not exposed to the user so it won't be a breaking change).

> I know that it's not dynamic then

It is dynamic, now we can reconfigure MeshMetric dynamically.

Reply:
I think @jakubdyszkiewicz was mentioning that: https://envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#envoy-v3-api-msg-config-metrics-v3-statssink  is dynamically configurable in envoy. 

Currently we have:

```json
        "statsConfig": {
          "statsTags": [
            {
              "tagName": "name",
              "regex": "^grpc\\.((.+)\\.)"
            },
            {
              "tagName": "status",
              "regex": "^grpc.*streams_closed(_([0-9]+))"
            },
            {
              "tagName": "kafka_name",
              "regex": "^kafka(\\.(\\S*[0-9]))\\."
            },
            {
              "tagName": "kafka_type",
              "regex": "^kafka\\..*\\.(.*?(?=_duration|$))"
            },
            {
              "tagName": "worker",
              "regex": "(worker_([0-9]+)\\.)"
            },
            {
              "tagName": "listener",
              "regex": "((.+?)\\.)rbac\\."
            }
          ]
        },
```

Reply:
Yeah so we could use `stats_matcher` but we need to be careful about https://github.com/envoyproxy/envoy/issues/8771